tools/python/util/onnx_model_utils.py

```diff
@@ -0,0 +1,311 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import logging
+import onnx
+import onnxruntime as ort
+import pathlib
+
+from onnx import version_converter
+
+
+def iterate_graph_per_node_func(graph, per_node_func, **func_args):
+    '''
+    Iterate the graph including subgraphs calling the per_node_func for each node.
+    :param graph: Graph to iterate
+    :param per_node_func: Function to call for each node. Signature is fn(node: onnx:NodeProto, **kwargs)
+    :param func_args: The keyword args to pass through.
+    '''
+
+    for node in graph.node:
+        per_node_func(node, **func_args)
+        # recurse into subgraph for control flow nodes (Scan/Loop/If)
+        for attr in node.attribute:
+            if attr.HasField('g'):
+                iterate_graph_per_node_func(attr.g, per_node_func, **func_args)
+
+
+def iterate_graph_per_graph_func(graph, per_graph_func, **func_args):
+    '''
+    Iterate the graph including subgraphs calling the per_graph_func for each Graph.
+    :param graph: Graph to iterate
+    :param per_graph_func: Function to call for each graph. Signature is fn(graph: onnx:GraphProto, **kwargs)
+    :param func_args: The keyword args to pass through.
+    '''
+
+    per_graph_func(graph, **func_args)
+
+    for node in graph.node:
+        # recurse into subgraph for control flow nodes (Scan/Loop/If)
+        for attr in node.attribute:
+            if attr.HasField('g'):
+                iterate_graph_per_graph_func(attr.g, per_graph_func, **func_args)
+
+
+def get_opsets_imported(model: onnx.ModelProto):
+    '''
+    Get the opsets imported by the model
+    :param model: Model to check.
+    :return: Map of domain to opset.
+    '''
+    opsets = {}
+    for entry in model.opset_import:
+        # if empty it's ai.onnx
+        domain = entry.domain or 'ai.onnx'
+        opsets[domain] = entry.version
+
+    return opsets
+
+
+def update_onnx_opset(model_path: pathlib.Path, opset: int, out_path: pathlib.Path = None,
+                      logger: logging.Logger = None):
+    """
+    Helper to update the opset of a model using onnx version_converter. Target opset must be greater than current opset.
+    Model is saved to the original location with the '.onnx' extension replaced with '.opset<opset>.onnx'.
```

Comment:
>Model is saved to the original location with the '.onnx' extension replaced with '.opset<opset>.onnx'. [](http://example.com/codeflow?start=4&length=102)

didn't see where the ".opset<opset>.onnx" path was set. is this in addition to out_path?

Reply:
Outdated comment. I was auto-generating an output path but that didn't fit with the actual usage so changed to require it to be specified.